File: netlify/functions/lib/scrapers/knaben.js
/**
 * Knaben scraper - HTML scraping.
 */
const client = require('../client');
const { parseSeeders, parseSize, cleanName } = require('../utils');

const NAME = 'Knaben';
const BASE = 'https://knaben.eu';

async function search(query) {
  const url = `${BASE}/search/index.php?q=${encodeURIComponent(query)}&search=fast`;
  const $ = await client.getSoup(url);
  if (!$) return [];

  const results = [];
  $('table tr, div.result').each((_, row) => {
    const $row = $(row);
    const magnetTag = $row.find('a[href^="magnet:"]').first();
    if (!magnetTag.length) return;
    const magnet = magnetTag.attr('href');

    let name = '';
    const nameTag = $row.find('a.title, td.name a, a[href*="/torrent/"]').first();
    if (nameTag.length) {
      name = cleanName(nameTag.text());
    }
    if (!name) {
      const dnMatch = (magnet || '').match(/dn=([^&]+)/);
      if (dnMatch) name = decodeURIComponent(dnMatch[1]).replace(/\+/g, ' ');
    }
    if (!name) return;

    const text = $row.text();
    const sizeMatch = text.match(/([\d,.]+\s*(?:GB|MB|KB|TB|GiB|MiB))/i);
    const sizeText = sizeMatch ? sizeMatch[1] : '';

    let seeders = 0;
    const seedTag = $row.find('td.seed, span.seed').first();
    if (seedTag.length) seeders = parseSeeders(seedTag.text());

    results.push({
      name,
      size: sizeText || 'N/A',
      size_gb: parseSize(sizeText),
      seeders,
      magnet,
      provider: NAME,
    });
  });
  return results;
}

module.exports = { name: NAME, base_url: BASE, search };

File: scrapers/knaben.py
"""Knaben scraper - HTML scraping."""

from urllib.parse import quote
from scrapers.base import BaseScraper
from scrapers import client, utils


class Source(BaseScraper):
    name = 'Knaben'
    base_url = 'https://knaben.eu'

    def search(self, query):
        url = f'{self.base_url}/search/index.php?q={quote(query)}&search=fast'
        soup = client.get_soup(url)
        if not soup:
            return []
        results = []
        for row in soup.select('table tr, div.result'):
            # Try to find magnet link
            magnet_tag = row.select_one('a[href^="magnet:"]')
            if not magnet_tag:
                continue
            magnet = magnet_tag['href']
            # Name
            name_tag = row.select_one('a.title, td.name a, a[href*="/torrent/"]')
            name = utils.clean_name(name_tag.get_text()) if name_tag else ''
            if not name:
                # Try getting name from the magnet link dn parameter
                import re
                dn_match = re.search(r'dn=([^&]+)', magnet)
                if dn_match:
                    from urllib.parse import unquote
                    name = unquote(dn_match.group(1)).replace('+', ' ')
            if not name:
                continue
            # Extract size and seeders from row text
            text = row.get_text()
            import re
            size_match = re.search(r'([\d,.]+\s*(?:GB|MB|KB|TB|GiB|MiB))', text, re.I)
            size_text = size_match.group(1) if size_match else ''
            seeders = 0
            seed_tag = row.select_one('td.seed, span.seed')
            if seed_tag:
                seeders = utils.parse_seeders(seed_tag.get_text())
            results.append({
                'name': name,
                'size': size_text or 'N/A',
                'size_gb': utils.parse_size(size_text),
                'seeders': seeders,
                'magnet': magnet,
                'provider': self.name,
            })
        return results
